site/app.js

```diff
@@ -49,68 +64,135 @@ if (SP_PROTECTED === "false") {
       issuer: SP_ENTITY_ID,
       privateCert: spKey
     },
-    function (profile, done) {
+    function(profile, done) {
       done(null, {
         nameID: profile.nameID,
         nameIDFormat: profile.nameIDFormat,
         email: profile.email,
         date: Date.now()
       })
-    })
-
-  passport.use(samlStrategy);
-  passport.serializeUser(mapUser);
-  passport.deserializeUser(mapUser);
-
-  app.post('/sso/callback',
-    passport.authenticate('saml', {failureRedirect: '/', failureFlash: true}),
+    }
+  )
+
+  passport.use(samlStrategy)
+  passport.serializeUser(mapUser)
+  passport.deserializeUser(mapUser)
+  app.post(
+    "/sso/callback",
+    passport.authenticate("saml", { failureRedirect: "/", failureFlash: true }),
     callback
-  );
+  )
 
-  app.get('/login',
+  app.get(
+    "/login",
     preAuthenticate,
-    passport.authenticate('saml', {
-      failureRedirect: '/',
+    passport.authenticate("saml", {
+      failureRedirect: "/",
       failureFlash: true
     })
-  );
+  )
 
   // logout route is not an end-user flow. It is only added for test and troubleshooting purposes.
-  app.get('/logout', logout(IDP_LOGOUT))
-
-  app.get('/api/auth', requireAuthenticated, (request, response) => withoutEtag(response).send("AUTH"))
-
-  app.get('/saml/metadata', function (req, res) {
-    res.type('application/xml');
-    res.send(samlStrategy.generateServiceProviderMetadata(null, spCertificate));
+  app.get("/logout", logout(IDP_LOGOUT))
+
+  app.get("/api/auth", requireAuthenticated, (request, response) =>
+    withoutEtag(response).send("AUTH")
+  )
+
+  if (FEATURE_UPLOADCARE_IS_ENABLED === "true") {
+    const UPLOADCARE_SECRET_KEY = process.env.UPLOADCARE_SECRET_KEY
+    const UPLOADCARE_PUBLIC_KEY = process.env.GATSBY_UPLOADCARE_PUBLIC_KEY
+    const UPLOADCARE_STORAGE_ID = process.env.UPLOADCARE_STORAGE_ID
+    const UPLOAD_SIGNATURE_EXPIRY_SECONDS = process.env.UPLOAD_EXPIRY || 120
+    const ASSET_BUCKET_NAME = process.env.ASSET_BUCKET_NAME
+    const ASSET_BUCKET_REGION = process.env.ASSET_BUCKET_REGION || "eu-west-2"
+
+    const customStorage = new CustomStorage(
+      UPLOADCARE_STORAGE_ID,
+      ASSET_BUCKET_NAME,
+      ASSET_BUCKET_REGION
+    )
+
+    const api = new UploadcareApi(
+      UPLOADCARE_PUBLIC_KEY,
+      UPLOADCARE_SECRET_KEY,
+      customStorage
+    )
+
+    app.get(
+      "/api/uploadcare/token",
+      requireImageUploadAuthorized,
+      (request, response) => {
+        // TODO
+        // AUDIT: log the user id here
+        if (!UPLOADCARE_SECRET_KEY) {
+          console.error("Missing UPLOADCARE_SECRET_KEY, cannot generate secret")
+          response.status(500).send()
+        } else {
+          response.send(api.createSignature(UPLOAD_SIGNATURE_EXPIRY_SECONDS))
+        }
+      }
+    )
+
+    app.post(
+      "/api/uploadcare/copy",
+      requireImageUploadAuthorized,
+      (request, response) => {
+        // AUDIT: log the user id here
+        // TODO input validation
```

Comment:
See previous comment. Add an ONS ticket reference for outstanding work

Reply:
https://worthit.jira.com/browse/ONS-801